windows/src/buildtools/sentrytool/Keyman.System.AddPdbToPe.pas

```diff
@@ -0,0 +1,253 @@
+    Inc(section);
+  end;
+
+  debug_writeln(Format('End of section table is at %8.8x', [Integer(section) - Integer(s.Memory)]));
+  debug_writeln(Format('One more section takes us to %8.8x', [Integer(section) - Integer(s.Memory) + sizeof(_IMAGE_SECTION_HEADER)]));
+
+  if nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].VirtualAddress <> 0 then
+  begin
+    // Let's look up the debug data for our own sanity
+    error_writeln('This executable has a debug section. Not able to update this file.');
+    if AddPdbToPe_Debug then
+    begin
+      a := RVA2Offset(nh, nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].VirtualAddress);
+      debug_writeln(Format('RVA = %x addr = %x', [nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].VirtualAddress, a]));
+
+      if nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].Size < sizeof(TImageDebugDirectory) then
+        // Delphi compiler bug! should be size in bytes not entries
+        debugDirectoryEntryCount := nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].Size
+      else
+        debugDirectoryEntryCount := nh.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DEBUG].Size div sizeof(TImageDebugDirectory);
+      pdd := Pointer(PByte(s.Memory) + a);
+
+      for i := 1 to debugDirectoryEntryCount do
+      begin
+        debug_writeln(Format('pdd = %x %x ver = %x minorver = %x type = %x size = %d addr = %x ptr = %x', [
+          pdd.Characteristics, pdd.TimeDateStamp, pdd.MajorVersion, pdd.MinorVersion,
+          pdd._Type, pdd.SizeOfData, pdd.AddressOfRawData, pdd.PointerToRawData]));
+    //      FindDataDirectory(ms, nh, IMAGE_DIRECTORY_ENTRY_DEBUG);
+        if pdd._Type = IMAGE_DEBUG_TYPE_CODEVIEW then
+        begin
+          // Let's process the codeview section
+          pc := Pointer(PByte(s.Memory) + pdd.PointerToRawData);
+          debug_writeln('Signature: '+pc.CvSignature);
+          if pc.CvSignature = 'RSDS' then
+          begin
+            debug_writeln(Format('File Debug ID = %s-%d', [GuidToString(pc.Signature), pc.Age]));
+            debug_writeln(Format('File PDB = %s', [PAnsiChar(@pc.PdbFileName[0])]));
+          end;
+        end;
+        Inc(pdd);
+      end;
+    end;
+    Exit(False);
+  end;
+
+  // Ensure we have space for an additional section header
+  debug_writeln(Format('offset=%x', [DWORD(section) - DWORD(s.Memory)+sizeof(_IMAGE_SECTION_HEADER)]));
+  debug_writeln(Format('first=%x', [FirstSection.PointerToRawData]));
+  if DWORD(section) - DWORD(s.Memory) + sizeof(_IMAGE_SECTION_HEADER) > FirstSection.PointerToRawData then
+  begin
+    error_writeln('We don''t have space in the file header for a new section. Not going to try to relocate.');
+    Exit(False);
+  end;
+
+  // Calculate some new constants
+  SectionRawSize := nh.OptionalHeader.FileAlignment;
```

Comment:
Note for future self / other reviewers:

Everything before this within the`AddDebugID` function is sanity-checking and debug logging, to ensure we understand what we're actually doing and looking at before we perform the actual writes.  
- It might help to have a documentation comment state exactly this, just to make code lookup a bit simpler in the future.
- That said, it's not that hard to recognize with all the `debug_` and `error_` logging statements.

A short synopsis of the data gained amidst all the checking and logging to this point: we're writing a new "section" at the end of the executable file.  There's no manipulation of the original content, so at least this shouldn't break the executable's actual functionality if our data here isn't exactly right.

Reply:
That stated, the `SectionRawSize` bit here itself currently eludes me - that is, how we can safely set it here, without yet knowing/using the size of the debug dictionary.

Reply:
`SectionRawSize`: this is never going to be smaller than the size of the debug data, which is limited to `sizeof(_CV_INFO_PDB70) - 2 + Length('null.pdb') + 1` which is 35 bytes. The minimum `FileAlignment` is [512 bytes](https://docs.microsoft.com/en-us/windows/win32/debug/pe-format#file-headers).